src/service.py

```diff
@@ -49,15 +52,25 @@ def configure(self, snap_config: dict[str, Any]) -> None:
         self.snap_client.set(snap_config, typed=True)
 
 
-def snap_install(resource: str) -> SnapService:
-    """Install or refresh the snap, and return the snap service.
+def snap_install(resource: Optional[str]) -> SnapService:
+    """Install the snap, and return the snap service.
+
+    Before installing the snap, it will remove the canonical and the upstream snap to ensure that
+    the snap installed is not a resource. Removing the golang exporter also ensure that there is
+    no conflict using the one from Canonical.
 
     Raises an exception on error.
     """
+    snap.remove([SNAP_NAME, UPSTREAM_SNAP])
```

Comment:
**thought (non-blocking):** 

Instead of remove upstream snap for the user, maybe a blocked status is more suitable?

(Just want to make sure it doesn't have any side effort for the user, feel free to resolve it.)

Reply:
I don't mind this method - it's nice and seamless for the end user.

Reply:
But this logic means we'll be removing and re-installing the snap each time, which could lead to unecessary service restarts?

Reply:
I don't see any side effect of removing the upstream snap and installing the one that we are handling.

This will be called on charm upgrade or when installing so it's kind expected to restart the service. I don't see a big issue particularly .

Reply:
> I don't see any side effect of removing the upstream snap and installing the one that we are handling.

Agreed, but note that we're also removing the one we're handling here too.  Not sure if that's desired.

Reply:
One side effect of leaving it around is that they might use the same port for the exporter. Can we make the logic of detecting old exporter and removing old exporter in a single function, and add comment saying that this function can be removed in the future when we fully adopt the "new" exporter. I see this issue as a transient issue: nobody will ever notice that in the future

Reply:
Something like and we wouldn't need `UPSTREAM_NAME` constant
```python
def warn_and_remove_local_snap_resource():
  try:
    snap.remove(["golang-openstack-exporter"])
  except snap.SnapNotFoundError:
    pass
  if resource:
     logger.warning(
          "Local resources cannot be used. Installing the exporter %s from snap store",
          SNAP_NAME,
      ) 
```

Reply:
I did something similar to this, but we don't need to worry about this SnapNotFoundError because the [lib](https://github.com/canonical/openstack-exporter-operator/blob/main/lib/charms/operator_libs_linux/v2/snap.py#L573-L578) already handle when it's not present.